File: agents/sdk.py
#!/usr/bin/env python3
"""
pxOS Agent SDK

Provides a Python client for registering with the pxOS Agent Registry,
sending heartbeats, and reporting metrics via the time-series store.

Usage:
    from sdk import AgentSDK

    sdk = AgentSDK(base_url="http://localhost:3839")
    sdk.register("my-agent", capabilities=["monitor"], config={"region": "us-east"})
    sdk.start_heartbeat(interval=30)

    # Later...
    sdk.report_metric("cpu", 0.75)
"""

import json
import threading
import urllib.request
import urllib.error
import time


class AgentSDK:
    """Client SDK for the pxOS Agent Registry API."""

    def __init__(self, base_url="http://localhost:3839"):
        """
        Initialize the SDK.

        Args:
            base_url: Base URL of the pxOS server.
        """
        self.base_url = base_url.rstrip("/")
        self.agent_id = None
        self.agent_data = None
        self._heartbeat_timer = None

    def _request(self, method, path, body=None):
        """
        Make an HTTP request to the server.

        Args:
            method: HTTP method (GET, POST, PUT, DELETE).
            path: URL path (e.g., /api/v1/agents).
            body: Optional dict body to send as JSON.

        Returns:
            Parsed JSON response body, or None on error.

        Raises:
            RuntimeError: If the server returns an error status.
        """
        url = f"{self.base_url}{path}"
        data = json.dumps(body).encode("utf-8") if body is not None else None

        req = urllib.request.Request(
            url,
            data=data,
            headers={"Content-Type": "application/json"},
            method=method,
        )

        try:
            with urllib.request.urlopen(req, timeout=10) as response:
                raw = response.read().decode("utf-8")
                if raw:
                    return json.loads(raw)
                return None
        except urllib.error.HTTPError as e:
            raw = e.read().decode("utf-8") if e.fp else ""
            try:
                detail = json.loads(raw)
            except (json.JSONDecodeError, ValueError):
                detail = {"error": raw}
            raise RuntimeError(
                f"HTTP {e.code} on {method} {path}: {detail}"
            ) from e
        except urllib.error.URLError as e:
            raise RuntimeError(f"Connection error on {method} {path}: {e}") from e

    def register(self, name, capabilities=None, config=None):
        """
        Register this agent with the server.

        POSTs to /api/v1/agents with the agent's name, capabilities, and config.
        On success, stores the returned agent id for subsequent calls.

        Args:
            name: Agent display name.
            capabilities: Optional list of capability strings.
            config: Optional dict of agent configuration.

        Returns:
            dict with the full agent data returned by the server.

        Raises:
            RuntimeError: If registration fails.
        """
        payload = {"name": name}
        if capabilities is not None:
            payload["capabilities"] = capabilities
        if config is not None:
            payload["config"] = config

        result = self._request("POST", "/api/v1/agents", payload)
        self.agent_id = result["id"]
        self.agent_data = result
        return result

    def heartbeat(self):
        """
        Send a heartbeat ping to the server for the registered agent.

        PUTs to /api/v1/agents/:id/heartbeat.

        Returns:
            True if heartbeat was accepted.

        Raises:
            RuntimeError: If the agent is not registered or the request fails.
        """
        if not self.agent_id:
            raise RuntimeError("Agent not registered. Call register() first.")
        self._request("PUT", f"/api/v1/agents/{self.agent_id}/heartbeat")
        return True

    def start_heartbeat(self, interval=30):
        """
        Start an automatic heartbeat loop on a background daemon thread.

        Args:
            interval: Seconds between heartbeat pings (default 30).

        Raises:
            RuntimeError: If the agent is not registered.
        """
        if not self.agent_id:
            raise RuntimeError("Agent not registered. Call register() first.")

        self.stop_heartbeat()

        def _loop():
            while self._heartbeat_timer is not None:
                try:
                    self.heartbeat()
                except Exception:
                    pass  # Heartbeat failures are non-fatal
                time.sleep(interval)

        self._heartbeat_timer = threading.Thread(target=_loop, daemon=True)
        self._heartbeat_timer.start()

    def stop_heartbeat(self):
        """Stop the automatic heartbeat loop."""
        self._heartbeat_timer = None

    def report_metric(self, key, value):
        """
        Report a metric value to the time-series store.

        POSTs to /api/v1/cells with an agent-scoped key (agent:{id}:{key}).
        The server's cell subscription automatically records it into the
        time-series store.

        Args:
            key: Metric name (e.g., "cpu", "memory").
            value: Metric value (number or string).

        Returns:
            dict with the server response.

        Raises:
            RuntimeError: If the agent is not registered or the request fails.
        """
        if not self.agent_id:
            raise RuntimeError("Agent not registered. Call register() first.")

        scoped_key = f"agent:{self.agent_id}:{key}"
        return self._request("POST", "/api/v1/cells", {scoped_key: value})

File: agents/sdk.js
// agents/sdk.js
// pxOS Agent SDK for JavaScript/Node.js
//
// Provides a client for registering with the pxOS Agent Registry,
// sending heartbeats, reporting metrics, and working with the task queue.

import { randomUUID } from 'node:crypto';

export class AgentSDK {
    /**
     * @param {object} [options]
     * @param {string} [options.baseUrl='http://localhost:3839']
     * @param {typeof fetch} [options.fetch] - Custom fetch implementation (for testing).
     */
    constructor(options = {}) {
        this.baseUrl = (options.baseUrl || 'http://localhost:3839').replace(/\/+$/, '');
        this._fetch = options.fetch || globalThis.fetch;
        this.agentId = null;
        this.agentData = null;
    }

    /**
     * Make an HTTP request to the pxOS server.
     * @param {string} method
     * @param {string} path
     * @param {object} [body]
     * @returns {Promise<object|null>}
     */
    async _request(method, path, body) {
        const url = `${this.baseUrl}${path}`;
        const opts = {
            method,
            headers: { 'Content-Type': 'application/json' },
        };
        if (body !== undefined) {
            opts.body = JSON.stringify(body);
        }

        const res = await this._fetch(url, opts);
        const text = await res.text();
        let data = null;
        try { data = JSON.parse(text); } catch {}

        if (!res.ok) {
            const detail = data || { error: text };
            throw new Error(`HTTP ${res.status} on ${method} ${path}: ${JSON.stringify(detail)}`);
        }
        return data;
    }

    /**
     * Register this agent with the server.
     * @param {string} name
     * @param {object} [options]
     * @param {string[]} [options.capabilities]
     * @param {object} [options.config]
     * @returns {Promise<object>}
     */
    async register(name, options = {}) {
        const payload = { name };
        if (options.capabilities) payload.capabilities = options.capabilities;
        if (options.config) payload.config = options.config;

        const result = await this._request('POST', '/api/v1/agents', payload);
        this.agentId = result.id;
        this.agentData = result;
        return result;
    }

    /**
     * Send a heartbeat for the registered agent.
     * @returns {Promise<boolean>}
     */
    async heartbeat() {
        this._requireRegistered();
        await this._request('PUT', `/api/v1/agents/${this.agentId}/heartbeat`);
        return true;
    }

    /**
     * Report a metric value.
     * @param {string} key
     * @param {number|string} value
     * @returns {Promise<object>}
     */
    async reportMetric(key, value) {
        this._requireRegistered();
        const scopedKey = `agent:${this.agentId}:${key}`;
        return this._request('POST', '/api/v1/cells', { [scopedKey]: value });
    }

    /**
     * Claim the highest-priority pending task.
     * GETs pending tasks, then PUTs claim on the first one.
     * @returns {Promise<object|null>} The claimed task, or null if none available.
     */
    async claimTask() {
        this._requireRegistered();
        const tasks = await this._request('GET', '/api/v1/tasks?status=pending');
        if (!tasks || tasks.length === 0) return null;

        const taskId = tasks[0].id;
        return this._request('PUT', `/api/v1/tasks/${taskId}/claim`, {
            agentId: this.agentId,
        });
    }

    /**
     * Mark a task as completed.
     * @param {string} taskId
     * @param {*} [result]
     * @returns {Promise<object>}
     */
    async completeTask(taskId, result) {
        return this._request('PUT', `/api/v1/tasks/${taskId}/complete`, { result });
    }

    /**
     * Mark a task as failed.
     * @param {string} taskId
     * @param {string} error
     * @returns {Promise<object>}
     */
    async failTask(taskId, error) {
        return this._request('PUT', `/api/v1/tasks/${taskId}/fail`, { error });
    }

    /**
     * Create a new task in the queue.
     * @param {object} payload
     * @param {number} [priority]
     * @returns {Promise<object>}
     */
    async createTask(payload, priority) {
        const body = { payload };
        if (priority !== undefined) body.priority = priority;
        return this._request('POST', '/api/v1/tasks', body);
    }

    _requireRegistered() {
        if (!this.agentId) {
            throw new Error('Agent not registered. Call register() first.');
        }
    }
}
